### ISSUE TITLE
[critical bug] Enabling cache reuse leads to logit violations and tokens going out of range.

### ISSUE THREAD

**Issue by akhoroshev** [NVIDIA/TensorRT-LLM#2217]
Affected versions: 78f5c2936b7bdaa56859075a3f2fcc5a63952134, 5fa9436e17c2f9aeace070f49aa645d2577f676b
Affected APIs: GptManager, Executor
Hardware: H100

```
python examples/llama/convert_checkpoint.py --model_dir $hf_checkpoint --output_dir 1-gpu-tmp


trtllm-build --checkpoint_dir 1-gpu-tmp --output_dir 1-gpu --max_input_len 31744 --max_seq_len 32768 --max_num_tokens 32768 --max_batch_size 256 --gemm_plugin  float16 --use_paged_context_fmha enable
```

```
cat 1-gpu-tmp/config.json
{
    "mlp_bias": false,
    "attn_bias": false,
    "rotary_base": 300000,
    "rotary_scaling": null,
    "residual_mlp": false,
    "disable_weight_only_quant_plugin": false,
    "moe": {
        "num_experts": 0,
        "top_k": 0,
        "normalization_mode": null,
        "tp_mode": 0
    },
    "architecture": "LlamaForCausalLM",
    "dtype": "float16",
    "vocab_size": 42016,
    "hidden_size": 6656,
    "num_hidden_layers": 60,
    "num_attention_heads": 64,
    "hidden_act": "silu",
    "logits_dtype": "float32",
    "norm_epsilon": 1e-06,
    "position_embedding_type": "rope_gpt_neox",
    "max_position_embeddings": 32768,
    "num_key_value_heads": 16,
    "intermediate_size": 17920,
    "mapping": {
        "world_size": 1,
        "gpus_per_node": 8,
        "tp_size": 1,
        "pp_size": 1,
        "moe_tp_size": 1,
        "moe_ep_size": 1
    },
    "quantization": {
        "quant_algo": null,
        "kv_cache_quant_algo": null,
        "group_size": 128,
        "smoothquant_val": null,
        "clamp_val": null,
        "has_zero_point": false,
        "pre_quant_scale": false,
        "exclude_modules": null
    },
    "use_parallel_embedding": false,
    "embedding_sharding_dim": 0,
    "share_embedding_table": false,
    "head_size": 128,
    "qk_layernorm": false
}
```

While doing load testing server generates bad tokens (out of vocab range) and finally **segfaults**. `enableBlockReuse` is `true`.
top_k = 0, top_p = 0, stream = true

```
{"asctime":"2024-09-05 14:46:07,436502","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F40937A0A00","thread_id":"0x00007F4169DFF400","text":"token is out of range: 42015"}
{"asctime":"2024-09-05 14:46:07,436573","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F40C2215500","thread_id":"0x00007F417FBFF400","text":"token is out of range: 42015"}
{"asctime":"2024-09-05 14:46:07,436598","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F4180058600","thread_id":"0x00007F411FDFE400","text":"token is out of range: 42015"}
{"asctime":"2024-09-05 14:46:07,436661","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F40EBA47C00","thread_id":"0x00007F418A7FF400","text":"token is out of range: 42015"}
{"asctime":"2024-09-05 14:46:07,461130","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F40E6A95300","thread_id":"0x00007F4148DFF400","text":"token is out of range: 758775389"}
{"asctime":"2024-09-05 14:46:07,461136","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F413F08CF00","thread_id":"0x00007F41521FE400","text":"token is out of range: 798136723"}
{"asctime":"2024-09-05 14:46:07,461150","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F40BA059B00","thread_id":"0x00007F412D1FF400","text":"token is out of range: 42015"}
{"asctime":"2024-09-05 14:46:07,461163","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F40BB464100","thread_id":"0x00007F41521FE400","text":"token is out of range: 806487578"}
{"asctime":"2024-09-05 14:46:07,492488","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F40E6A95300","thread_id":"0x00007F41143FD400","text":"token is out of range: 763378200"}
{"asctime":"2024-09-05 14:46:07,492514","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F413F08CF00","thread_id":"0x00007F41423FE400","text":"token is out of range: 798136723"}
{"asctime":"2024-09-05 14:46:07,492525","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F40BB464100","thread_id":"0x00007F41A1FFF400","text":"token is out of range: 806487578"}
{"asctime":"2024-09-05 14:46:07,492531","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F40BA059B00","thread_id":"0x00007F41379FF400","text":"token is out of range: 42015"}
{"asctime":"2024-09-05 14:46:07,592616","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F40E6A95300","thread_id":"0x00007F4140FFF400","text":"token is out of range: 743574649"}
{"asctime":"2024-09-05 14:46:07,592647","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F40E6A95300","thread_id":"0x00007F4140FFF400","text":"token is out of range: 722043104"}
{"asctime":"2024-09-05 14:46:07,592648","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F40E6A95300","thread_id":"0x00007F4140FFF400","text":"token is out of range: 776627824"}
{"asctime":"2024-09-05 14:46:07,592661","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F40BB464100","thread_id":"0x00007F4140FFF400","text":"token is out of range: 806487578"}
{"asctime":"2024-09-05 14:46:07,592650","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F413F08CF00","thread_id":"0x00007F419CBFF400","text":"token is out of range: 798136723"}
{"asctime":"2024-09-05 14:46:07,592665","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F40BB464100","thread_id":"0x00007F4140FFF400","text":"token is out of range: 806487578"}
{"asctime":"2024-09-05 14:46:07,592669","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F40BB464100","thread_id":"0x00007F4140FFF400","text":"token is out of range: 806487578"}
{"asctime":"2024-09-05 14:46:07,592674","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F413F08CF00","thread_id":"0x00007F419CBFF400","text":"token is out of range: 798136723"}
{"asctime":"2024-09-05 14:46:07,592675","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F413F08CF00","thread_id":"0x00007F419CBFF400","text":"token is out of range: 798136723"}
{"asctime":"2024-09-05 14:46:07,592658","level":"WARNING","module":"Feed ( src/lib/pipeline/src/stream_tokenizer.cpp:36 ) ","task_id":"7F40BA059B00","thread_id":"0x00007F41705FF400","text":"token is out of range: 42015"}
```
NOTE: 42015 is out of range in terms of real tokenizer size (which is equal to 42002).

But token ids 758775389, 806487578, etc are absolutely abnormal.

All works fine if I set `enableBlockReuse` to `false`.

**Comment by akhoroshev** [comment]
Load testing scenario is:

Concurrent shot (up to 200 users) with 3 kind of requests:

```
1. system_prompt + user_message
2. system_prompt + (user_message * 10)
3. system_prompt + (user_message * 60)
```

system_prompt and user_message are constants.
system_prompt size is 54, user_message size is 311.

**Comment by akhoroshev** [comment]
I found this line https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/tensorrt_llm/kernels/samplingTopKKernels.cu#L227
`If sId is -1 here we force output token to the last from vocabulary to get vivid indicator of smth going wrong for the debug` and printed `logProbs`. They were all `nan`.

**Comment by akhoroshev** [comment]
@byshiue

**Comment by akhoroshev** [comment]
I guess you have a race condition problem.

I made load testing and trace situations when logits are going violations. I came across an interesting case:

```
1. 13:11:05 -- accepted req_0 (content: user_message)
2. 13:11:10 -- accepted req_1 (content: user_message * 60)
3. 13:11:21 -- successful processed req_0
4. 13:12:07 -- received first broken token for req_1 (logits are NaN)
```

Looks like using an unfinished kv cache from req_0 for req_1

**Comment by akhoroshev** [comment]
It reproduces with disabled `enableBlockReuse` option.

https://github.com/NVIDIA/TensorRT-LLM/issues/2255
